Make the build config more developer-friendly.

juliendehos opened the pull request:
Hi,
The miso-isomorphic-example project is great for building final applications but I think it could be improved for developing them (i.e. incremental build & separate testing). Here is a proposal for such improvement. I know these changes are quite questionable so feel free to modify or ignore them.


The aim of this commit is to make the nix config usable for both the build (nix-build) and the development (nix-shell + cabal). Now, you can "just nix-build" as before but also "run a nix-shell" in the subdirectories to incrementally build and test the subprojects (common, client, server). Important changes:
- add some cabal files & default.nix in the subdirectories
- use callCabal2nix to automatically generate nix configs from cabal files
- rename the "shared" directory to "common" (to avoid potential confusion with the "shared" keyword in the cabal files)
- add a nixpkgs.nix file for building a compatible nixpkgs environment
- add a section in the README.md explaining how to use the development config

Commit:
Make the build config more developer-friendly.

The aim of this commit is to make the nix config usable for both the
build (nix-build) and the development (nix-shell + cabal). Now, you can
"just nix-build" as before but also "run a nix-shell" in the
subdirectories to incrementally build and test the subprojects (common,
client, server).

Important changes:

- add some cabal files & default.nix in the subdirectories

- use callCabal2nix to automatically generate nix configs from cabal
files

- rename the "shared" directory to "common" (to avoid potential
confusion with the "shared" keyword in the cabal files)

- add a nixpkgs.nix file for building a compatible nixpkgs environment

- add a section in the README.md explaining how to use the development
config

FPtje commented on README.md:
````diff
@@ -47,3 +47,50 @@ cd $(nix-build) && bin/server
+```bash
+cd client
+nix-shell
+cabal new-build
````
Is there any specific reason to use `new-build`? I'm not too familiar with it, but reading the [wiki](https://www.haskell.org/cabal/users-guide/nix-local-build-overview.html) it doesn't seem to have a significant improvement if you already use nix. Correct me if I'm wrong.

juliendehos replied:
You're right, `new-build` is not necessary. I used the `cabal.project` file + `new-build` to specify the ghcjs compiler but it can be done using a `cabal.config` file + traditional cabal commands.

FPtje commented on README.md:
````diff
@@ -47,3 +47,50 @@ cd $(nix-build) && bin/server
+exit
+```
+
+You need to initialize the server before running `cabal run server`:
````
You might want to rephrase `initialize the server` to something more specific, like `make the client available to the server`.

juliendehos replied:
Yes. Fixed.

FPtje commented on client/cabal.project:
```diff
@@ -0,0 +1,4 @@
+compiler: ghcjs
```
Shouldn't this file be on the top level, and describe both server and client?
On another note, what should the workflow be regarding nix shell? The dependency sets of the client and server are very much different because of the difference between ghc and ghcjs. How should one start a nix shell that allows them to build the entire project?

juliendehos replied:
In this config, the sub-projects (common, client, server) are quite independent. Regarding nix-shell, the workflow should be: 
- go in a sub-directory
- run a `nix-shell`
- write code and run `cabal build`
- exit the nix-shell and apply the same workflow in another sub-directory

That said, I wrote an alternative config (https://github.com/juliendehos/miso-isomorphic-example/tree/config2) with the following workflow:
- run `nix-shell -A client` in the main directory
- write code and run `cabal --builddir=dist-client --config-file=client.config build`
- exit + same with the similar server config

This may be simpler but I think the cabal commands are too complicated. This could be solved by adding an alias in the cabal2nix config (e.g. `alias cabal-server="cabal --builddir=dist-server --config-file=server.config"`) but I don't how to do that.

FPtje commented on client/default.nix:
```diff
@@ -1,11 +1,22 @@
+
+let
+
+  _client = pkgs.haskell.packages.ghcjs.callCabal2nix "client" ./. {
```
Why do these variables start with underscores? This is very unusual in let bindings AFAIK.

juliendehos replied:
It's a personal habit to distinguish local variables but it's indeed unusual. Fixed.

FPtje commented on common/common-test.hs:
```diff
@@ -0,0 +1,5 @@
+import Common (initialModel, homeLink)
+
+main :: IO ()
+main = print $ initialModel homeLink
```
I understand that adding a test case makes the example more "complete", but I think I prefer being minimal over being complete. Since adding a test case to a project shouldn't be that difficult, I reckon it's best to leave it out.

juliendehos replied:
Ok. Fixed.

FPtje commented on default.nix:
```diff
@@ -1,62 +1,14 @@
-  };
+  pkgs = import ./nixpkgs.nix ;
+  server = pkgs.callPackage ./server/default.nix {};
+  client = pkgs.callPackage ./client/default.nix {};
```
This is *really* good. Separation of concerns and readable code. Good job.

juliendehos replied:
Thanks.

FPtje commented on client/default.nix:
```diff
@@ -1,11 +1,22 @@
+    src = ./.;
+    installPhase = ''
+      mkdir -p $out/static
+      ${pkgs.closurecompiler}/bin/closure-compiler ${_client}/bin/client.jsexe/all.js > $out/static/all.js
```
Making a separate package to run the closure compiler is actually a nifty idea. It might be a good idea, though, to override the original package. That way, you should end up with only *one* store path for the package.

juliendehos replied:
It's a good idea but I don't know how to override the derivation produced by cabal2nix. Another "solution" is to run the closure compiler in the main `default.nix` only.

FPtje reviewed (changes requested):
Making the packages independently buildable is a great idea. It also makes the example more usable as a start of a real project. See the review for some questions and comments.

Commit:
make the requested changes:

- change `cabal new-build` to `cabal build`

- fix and update README.md

- fix variable names in nix files (remove the underscores)

- remove the test app in the `common` project

juliendehos commented:
Thanks for your comments. Please see my responses and also consider this alternative config: https://github.com/juliendehos/miso-isomorphic-example/tree/config2. Personally, I have no particular preference for one over the other.

FPtje commented:
I prefer this config, as the `cabal` commands are simpler this way. Thanks for the contribution, I really appreciate making this example easier to use and build upon.
